File: src/main/java/org/antop/java/chapter4/CaseInsensitiveFileComparator.java
package org.antop.java.chapter4;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;

public class CaseInsensitiveFileComparator implements Comparator<Path> {
    @Override
    public int compare(Path o1, Path o2) {
        var filename1 = o1.getFileName().toString();
        var filename2 = o2.getFileName().toString();
        return String.CASE_INSENSITIVE_ORDER.compare(filename1, filename2);
    }

    private CaseInsensitiveFileComparator() {
    }

    public static CaseInsensitiveFileComparator INSTANCE = new CaseInsensitiveFileComparator();

    public static void main(String[] args) {
        System.out.println(
                CaseInsensitiveFileComparator.INSTANCE.compare(
                        Paths.get("/User"),
                        Paths.get("/user")
                )
        );

        var paths = new ArrayList<>(Arrays.asList(Paths.get("/Z"), Paths.get("/a")));
        paths.sort(CaseInsensitiveFileComparator.INSTANCE);
        System.out.println(paths);
    }

}

File: src/main/kotlin/org/antop/kotlin/chapter4/CaseInsensitiveFileComparator.kt
package org.antop.kotlin.chapter4

import java.nio.file.Path
import java.nio.file.Paths

object CaseInsensitiveFileComparator : Comparator<Path> {
    override fun compare(o1: Path, o2: Path): Int {
        val filename1 = o1.fileName.toString()
        val filename2 = o2.fileName.toString()
        return filename1.compareTo(filename2, ignoreCase = true)
    }
}

fun main() {
    println(
        CaseInsensitiveFileComparator.compare(
            Paths.get("/User"),
            Paths.get("/user")
        )
    )

    val paths = listOf(Paths.get("/Z"), Paths.get("/a"))
    println(paths.sortedWith(CaseInsensitiveFileComparator))
}
